### PR TITLE
Fix #3029: Use correct server class when distributing Play applications

### LINKED ISSUE #3029: playBinary - gradle 4.2 play 2.6 - using the wrong main class
`gradle dist` generates `playBinary `, which contains at line 165
```
eval set -- $DEFAULT_JVM_OPTS $JAVA_OPTS $PLAY_BINARY_OPTS -classpath "\"$CLASSPATH\"" play.core.server.NettyServer "$APP_ARGS"
```  
which causes the following error when run  `playBinary `:  
`Could not find or load main class play.core.server.NettyServer    
`    
  
  
Looks like the correct class should be: `play.core.server.ProdServerStart`  

Should this just a be simple fix?
Thanks  

-cl

### PR DESCRIPTION
**Pull request by marcospereira**
### Context

This fixes #3029 where the wrong main class is used when running dist tasks. The full qualified class name for Play 2.6 is `"play.core.server.ProdServerStart"`

### Contributor Checklist

- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist

- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by blindpirate** [comment]
Thanks for your excellent work! LGTM, but please give me some time to validate it. @big-guy I'll appreciate it if you can give us more suggestions.

**Comment by marcospereira** [comment]
@blindpirate thanks for looking. Not trying to rush your guys here, so take your time. My goal here is to fix these small problems so that we can have full support of Play 2.6 and then add Gradle's `build.gradle` file to our example projects and templates (eventually documentation too).

Besides that, I really appreciate that we can work together to evolve Gradle+Play dev experience. ;-)

File: subprojects/platform-play/src/main/java/org/gradle/play/plugins/PlayDistributionPlugin.java [hunk 1]
```diff
@@ -168,6 +170,22 @@ public void execute(CreateStartScripts createStartScripts) {
...
+        switch (PlayMajorVersion.forPlatform(playPlatform)) {
+            case PLAY_2_2_X:
+            case PLAY_2_3_X:
+            case PLAY_2_4_X:
+            case PLAY_2_5_X:
+                return "play.core.server.NettyServer";
+            case PLAY_2_6_X:
+                return "play.core.server.ProdServerStart";
+            default:
+                throw new RuntimeException("Could not create Twirl compile spec for Play version: " + playVersion);
```
Comment (big-guy) [inline subprojects/platform-play/src/main/java/org/gradle/play/plugins/PlayDistributionPlugin.java]: This should probably say something like, `"Could not determine main class for Play version: " + playVersion`

File: subprojects/platform-play/src/main/java/org/gradle/play/plugins/PlayDistributionPlugin.java [hunk 2]
```diff
@@ -168,6 +170,22 @@ public void execute(CreateStartScripts createStartScripts) {
...
+        PlayPlatform playPlatform = distribution.getBinary().getTargetPlatform();
+        String playVersion = playPlatform.getPlayVersion();
+        switch (PlayMajorVersion.forPlatform(playPlatform)) {
+            case PLAY_2_2_X:
+            case PLAY_2_3_X:
+            case PLAY_2_4_X:
+            case PLAY_2_5_X:
+                return "play.core.server.NettyServer";
+            case PLAY_2_6_X:
+                return "play.core.server.ProdServerStart";
```
Comment (big-guy) [inline subprojects/platform-play/src/main/java/org/gradle/play/plugins/PlayDistributionPlugin.java]: I think this has actually been deprecated for a few releases (2.4.x?).  Could we try using the non-deprecated version for 2.4+?

Comment (marcospereira) [reply subprojects/platform-play/src/main/java/org/gradle/play/plugins/PlayDistributionPlugin.java]: You are right:

https://github.com/playframework/playframework/blob/2.4.x/framework/src/play-netty-server/src/main/scala/play/core/server/NettyServer.scala#L230-L233

Will use `ProdServerStart` for 2.4 and 2.5 too.

File: subprojects/platform-play/src/main/java/org/gradle/play/plugins/PlayDistributionPlugin.java (hunk 1 above)
Comment (marcospereira) [reply subprojects/platform-play/src/main/java/org/gradle/play/plugins/PlayDistributionPlugin.java]: Bad copy and pasta. :-P

Will fix that too.

**Comment by marcospereira** [comment]
@big-guy thanks for helping here.

Please, review again.

File: subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayDistributionPluginTest.groovy [hunk 3]
```diff
@@ -157,6 +159,13 @@ class PlayDistributionPluginTest extends Specification {
                 }
             })
         }
+
+        where:
+        playVersion | mainClass
+        "2.3.10"    | "play.core.server.NettyServer"
+        "2.4.11"    | "play.core.server.NettyServer"
+        "2.5.18"    | "play.core.server.NettyServer"
```
Comment (benmccann) [inline subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayDistributionPluginTest.groovy]: If I'm reading the code in `PlayDistributionPlugin.java` correctly then `NettyServer` is expected for 2.2 and 2.3 and `ProdServerStart` for 2.4+. I'm not sure whether the code or the test is correct, but they don't seem to match?

Comment (blindpirate) [reply subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayDistributionPluginTest.groovy]: Yes you're right, I guess this is just a typo. I've fixed them, thanks for pointing out!

**Comment by marcospereira** [comment]
@big-guy and @blindpirate, some questions: are these fixes ok to be backported to Gradle 4.3? I don't know how this works for Gradle (there is no information in contributing guide). As far as I can tell, these changes are compatible with what we have in 4.3 so they should be ok to backport.

Should I submit new PR targeting a specific branch or are the changes cherry-picked manually when releasing? Is it possible to add then to the next dot release?

**Comment by marcospereira** [comment]
By the way, I've noticed that there is a "4.4 RC1" milestone and that #3018 was added to this milestone. I think this PR could be part of "4.4 RC1" too.

**Pull request merged**